### PR TITLE
Event mesh mode

### PR DESCRIPTION
**Pull request by muralov**
**Description**

Create EPP deployment for EventMesh mode

Changes proposed in this pull request:

- Implementation
- Tests

**Related issue(s)**
#40

### CODE REVIEW

File: internal/controller/eventing/eventmesh.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	err = setUpEnvironmentForEventMesh(secretForPublisher)
+	if err != nil {
+		// backendStatus.SetSubscriptionControllerReadyCondition(false, eventingv1alpha1.ConditionReasonControllerStartFailed, err.Error())
+		// if updateErr := r.syncBackendStatus(ctx, backendStatus, nil); updateErr != nil {
+		// 	return ctrl.Result{},fmt.Errorf(err, "failed to update status while setting up environment variables for BEB controller")
+		// }
+		return fmt.Errorf("failed to setup environment variables for EventMesh controller: %v", err)
+	}
+
+	// TODO: start subscription controller
```
Comment (marcobebway) [inline internal/controller/eventing/eventmesh.go:45]: Do we still need this TODO? Is there a follow up ticket for it? Please link it here if any.

File: internal/controller/eventing/eventmesh.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		return err
+	}
+
+	// Set environment with secrets for BEB subscription controller
+	err = setUpEnvironmentForEventMesh(secretForPublisher)
+	if err != nil {
+		// backendStatus.SetSubscriptionControllerReadyCondition(false, eventingv1alpha1.ConditionReasonControllerStartFailed, err.Error())
+		// if updateErr := r.syncBackendStatus(ctx, backendStatus, nil); updateErr != nil {
+		// 	return ctrl.Result{},fmt.Errorf(err, "failed to update status while setting up environment variables for BEB controller")
+		// }
```
Comment (marcobebway) [inline internal/controller/eventing/eventmesh.go:41]: Please remove commended code.

File: internal/controller/eventing/eventmesh_test.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+		return secret
+	}
+
+	testCases := []struct {
+		name           string
+		messagingData  []byte
+		namespaceData  []byte
+		expectedSecret corev1.Secret
+		expectedError  error
```
Comment (marcobebway) [inline internal/controller/eventing/eventmesh_test.go:41]: For consistency with tests in this repo, please keep the pattern:
- givenABC
- wantXYZ

File: internal/controller/eventing/integrationtests/controller/integration_test.go [hunk 4]
```diff
@@ -153,32 +157,103 @@ func Test_CreateEventingCR(t *testing.T) {
...
+				utils.WithEventingPublisherData(2, 2, "199m", "99Mi", "399m", "199Mi"),
+				utils.WithEventingEventTypePrefix("test-prefix"),
+			),
+			wantMatches: gomega.And(
+				matchers.HaveStatusProcessing(),
+				//matchers.HaveNATSAvailableConditionAvailable(),
+				matchers.HavePublisherProxyReadyConditionProcessing(),
+				matchers.HaveFinalizer(),
+			),
+		},
```
Comment (marcobebway) [inline internal/controller/eventing/integrationtests/controller/integration_test.go:214]: Should there be one more case for `Eventing CR should have error state ...`?

File: pkg/k8s/client_test.go [hunk 5]
```diff
@@ -148,7 +150,78 @@ func Test_DeleteDeployment(t *testing.T) {
...
+		})
+	}
+}
+
+func Test_GetSecret(t *testing.T) {
+	t.Parallel()
+	// Define test cases as a table.
+	testCases := []struct {
+		name              string
+		namespacedName    string
```
Comment (marcobebway) [inline pkg/k8s/client_test.go:163]: Please follow the give/want naming pattern here.

File: internal/controller/eventing/controller.go [hunk 6]
```diff
@@ -267,8 +277,18 @@ func (r *Reconciler) handlePublisherProxy(
 }
 
 func (r *Reconciler) reconcileEventMeshBackend(ctx context.Context, eventing *eventingv1alpha1.Eventing, log *zap.SugaredLogger) (ctrl.Result, error) {
-	// TODO: Implement me.
-	return ctrl.Result{}, nil
+	// Start the EventMesh subscription controller
+	err := r.startEventMeshSubscriptionController(ctx, eventing)
+	if err != nil {
+		// TODO: implement status handling
```
[the remark this answers is not in the document]
Comment (muralov) [reply internal/controller/eventing/controller.go:283]: there is no follow-up ticket, it is part of this ticket. I just created a PR for EPP creation first. Our small PR principle ;)

File: internal/controller/eventing/integrationtests/controller/integration_test.go (hunk 4 above)
Comment (muralov) [reply internal/controller/eventing/integrationtests/controller/integration_test.go:214]: I've really tried, but it is quite hard to get error case here as we don't mock k8s api server here.

**Review by marcobebway — APPROVED** [review]

**Pull request merged**